mailbox.hpp

```diff
@@ -0,0 +1,118 @@
+#include <vector>
+#include <queue>
+#include <memory>
+#include <new>
+#include <set>
+#include <mutex>
+#include <cstring>
+#include <cassert>
+#include <algorithm>
+#include <unordered_map>
+#include <condition_variable>
+
+#include "lender.hpp"
+
+
+template <typename T>
+mailbox<T>::mailbox(const std::vector<size_t> s, int max, int reads) : 
+    recycle_memory<T>::recycle_memory(s, max) 
+{
+    box = std::unordered_map<int, std::shared_ptr<map_value>>();
+    max_read = reads;
+}
+
+template <typename T>
+bool mailbox<T>::contains_key(int idx) { return box.find(idx) != box.end(); }
+
+template <typename T>
+auto mailbox<T>::fill() -> buffer_ptr<T> {
+    // take from free and give to caller
+
+    std::unique_lock<std::mutex> lock(this->free_mutex);
+    while(this->free_q.empty()) {  
+        this->free_variable.wait(lock);
+    }
+
+    // take from free list
+    T* ptr = this->free_q.front();
+    this->free_q.pop_front();
+
+    // make reuseable_buffer for the buffer
+    auto sp = buffer_ptr<T>(ptr, *this);
+    return sp;
+}
+
+template <typename T>
+void mailbox<T>::queue (int key, buffer_ptr<T> ptr) {
+    // add new entry or fill in old
+    std::unique_lock<std::mutex> guard(box_lock);
+
+    if (contains_key(key)) {
+        // need to take mutex and inform cd
+        std::unique_lock<std::mutex> lock(box[key]->val_lock);
+        box[key]->value = ptr;
+        box[key]->count = max_read;
+        box_cv.notify_one();
+
+    } else {
+        std::shared_ptr<map_value> val = std::make_shared<map_value>();
+        std::unique_lock<std::mutex> lock(val->val_lock);
+
+        val->value = ptr;
+        val->count = max_read;
+        box.emplace(key, val);
+        
```

Comment:
No call to `box_cv.notify_one()` here? Also, although I may be mistaken, it looks like you are notifying threads waiting on the mailbox for the addition of (or change to) a `map_value` value in the mailbox. If that's the case, then you might consider the use of `notify_all()` instead of `notify_one()`, which could also lead to a simplification of the `mailbox::operate()` method (where it appears currently to be doing the job of notifying other waiters explicitly, one at a time).